tla/ccfraft.tla

```diff
@@ -121,98 +132,248 @@ VARIABLE configurations
 \* removed messages once received.
 VARIABLE messages
 
+\* Helper function for checking the type safety of log entries
+EntryTypeOK(entry) ==
+    /\ entry.term \in Nat \ {0}
+    /\ \/ /\ entry.contentType = TypeEntry
+          /\ entry.value \in Nat \ {0}
+       \/ /\ entry.contentType = TypeSignature
+          /\ entry.value = Nil
+       \/ /\ entry.contentType = TypeReconfiguration
+          /\ entry.value \subseteq Servers
+
+AppendEntriesRequestTypeOK(m) ==
+    /\ m.mtype = AppendEntriesRequest
+    /\ m.mprevLogIndex \in Nat
+    /\ m.mprevLogTerm \in Nat
+    /\ m.mcommitIndex \in Nat
+    /\ \A k \in DOMAIN m.mentries: EntryTypeOK(m.mentries[k])
+
+AppendEntriesResponseTypeOK(m) ==
+    /\ m.mtype = AppendEntriesResponse
+    /\ m.msuccess \in BOOLEAN
+    /\ m.mmatchIndex \in Nat
+
+RequestVoteRequestTypeOK(m) ==
+    /\ m.mtype = RequestVoteRequest
+    /\ m.mlastLogTerm \in Nat
+    /\ m.mlastLogIndex \in Nat
+
+RequestVoteResponseTypeOK(m) ==
+    /\ m.mtype = RequestVoteResponse
+    /\ m.mvoteGranted \in BOOLEAN
+
+NotifyCommitMessageTypeOK(m) ==
+    /\ m.mtype = NotifyCommitMessage
+    /\ m.mcommitIndex \in Nat
+
+MessagesTypeInv ==
+    \A m \in messages :
+        /\ m.msource \in Servers
+        /\ m.mdest \in Servers
+        /\ m.mterm \in Nat \ {0}
+        /\ \/ AppendEntriesRequestTypeOK(m)
+            \/ AppendEntriesResponseTypeOK(m)
+            \/ RequestVoteRequestTypeOK(m)
+            \/ RequestVoteResponseTypeOK(m)
+            \/ NotifyCommitMessageTypeOK(m)
+
 \* CCF: Keep track of each append entries message sent from each server to each other server
 \* and cap it to a maximum to constraint the state-space for model-checking.
 \* TLC: Finite state space.
 VARIABLE messagesSent
 
+MessagesSentTypeInv ==
+    \A i,j \in Servers : i /= j =>
+        \A k \in DOMAIN messagesSent[i][j] :
+            messagesSent[i][j][k] \in Nat \ {0}
+
 \* CCF: After reconfiguration, a RetiredLeader leader may need to notify servers
 \* of the current commit level to ensure that no deadlock is reached through
 \* leaving the network after retirement (as that would lead to endless leader
 \* re-elects and drop-outs until f is reached and network fails).
 VARIABLE commitsNotified
 
-messageVars == <<messages, messagesSent, commitsNotified>>
-----
+CommitsNotifiedTypeInv ==
+    \A i \in Servers :
+        /\ commitsNotified[i][1] \in Nat
+        /\ commitsNotified[i][2] \in Nat
+
+messageVars == <<
+    messages, 
+    messagesSent, 
+    commitsNotified
+>>
+
+MessageVarsTypeInv ==
+    /\ MessagesTypeInv
+    /\ MessagesSentTypeInv
+    /\ CommitsNotifiedTypeInv
+
+
+------------------------------------------------------------------------------
 \* The following variables are all per server (functions with domain Servers).
 
 \* The server's term number.
 VARIABLE currentTerm
 
+CurrentTermTypeInv ==
+    \A i \in Servers : currentTerm[i] \in Nat
```

Comment:
```suggestion
    \A i \in Servers : currentTerm[i] \in Nat \ {0, 1, 2}
```
if we want to match the implementation, which I think we do since we exclude 0 for seqno/index.

Reply:
Currently in the spec each node's `currentTerm` starts at 0 so this isn't true. 

https://github.com/microsoft/CCF/blob/9f8833d41abd47650b3f771f1b7e92fafedfa436/tla/ccfraft.tla#L392-L395

Not quite sure I understand what you mean about excluding 0 for seqno/index

Reply:
@heidihoward my bad, I confused `entry.value` with `index`. At line 194 though, we seemingly exclude 0 for term:

```
MessagesTypeInv ==
    \A m \in messages :
        /\ m.source \in Servers
        /\ m.dest \in Servers
        /\ m.term \in Nat \ {0}
``

Reply:
The existing trace validation scenarios confirm that there is never a message with term = 0.

Reply:
Since all nodes start in term 0 as followers, the first candidate is in term 1 so there should never be a message with term 0

Reply:
Maybe somewhere add `\A s \in Servers: currentTerm[s] = 0 => state[s] # Leader`?

Reply:
Done in 147fed0bb023e075550b37ff9cdbebdbf132554a